### DISCUSSION TITLE
Preventing negative variable values, or assigning range of possible values to variable

### DISCUSSION THREAD

**Discussion by EdSheltonUKAEA** [idaholab/moose#23210] [Q&A General] [answered]
Is there a way to prevent a Variable (or AuxVariable) from taking negative values when this would be physically meaningless?
More generally is it possible to assign a range of possible values to a variable?
There doesn't seem to be anything in Variable parameters which does this - unless I have missed it.
Is there a Kernel which can restrict the range of values of a variable?
I see there are these optional parameter for the Materials type ParsedMaterial (and ADParsedMaterial?):
```
tol_names
C++ Type:std::vector<std::string>
Controllable:No
Description:Vector of variable names to be protected from being 0 or 1 within a tolerance (needed for log(c) and log(1-c) terms)

tol_values
C++ Type:std::vector<double>
Controllable:No
Description:Vector of tolerance values for the variables in tol_names
```
It looks like these might help do what I want, but I'm not sure of the syntax of how to apply these.  Does anyone have any examples they can share?

**Comment by GiudGiud** [comment, marked as answer]
Hello

you can use the Bounds system to impose that. There are files under auxkernels

for auxiliary variables, auxkernels can be made to bound the variable any way you like, these restrictions will just need to appear in the main value routine of the auxkernel

Guillaume

**Reply by EdSheltonUKAEA** [reply]
That's brilliant.  Thanks so much.  This has done exactly what I wanted to stop variables taking negative values.  However, I also have at least one ADParsedMaterial that I need to prevent being negative.  Is there an equivalent way of doing this?
Is there a kernel which can effectively just assign the value of the material property to a variable, so that I can then use the Bounds system to stop that variable being negative?

**Reply by EdSheltonUKAEA** [reply]
Hold on, I think I might have just found it: 
https://mooseframework.inl.gov/source/kernels/MaterialPropertyValue.html

**Reply by GiudGiud** [reply]
With a parsedMaterial you can use an if(a>0, a, 0) syntax to force it to always be positive
